### PR TITLE
go@1.9 1.9.4 (new formula)

### PR DESCRIPTION
**Pull request by mkabischev**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by fxcoudert** [comment]
> Each major Go release is supported until there are two newer major releases. For example, Go 1.8 is supported until Go 1.10 is released, and Go 1.9 is supported until Go 1.11 is released.

**Comment by fxcoudert** [comment]
- Versioned formula should be, as far as possible, optionless.
- Should be keg-only, so as not to conflict with the main go formula.

**Comment by MikeMcQuaid** [comment]
> Versioned formula should be, as far as possible, optionless.

@fxcoudert This is still under discussion, FYI.

**Comment by mkabischev** [comment]
@fxcoudert I know about go release cycle. 1.9 will be supported for next 6 months until 1.11.

It's my first formula for homebrew, so can you tell what does
> Should be keg-only, so as not to conflict with the main go formula.
mean?

thanks.

File: Formula/go@1.9.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    sha256 "0573a8df33168977185aa44173305e5a0450f55213600e94541604b75d46dc06"
+
+    go_version = version.to_s.split(".")[0..1].join(".")
+    resource "gotools" do
+      url "https://go.googlesource.com/tools.git",
+          :branch => "release-branch.go#{go_version}"
+    end
+  end
+
+  bottle do
```
Comment (DomT4) [inline Formula/go@1.9.rb]: Don't add this yourself.

File: Formula/go@1.9.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    sha256 "f888bb85ad375f7e7ee683e1d575de4ec404a9a26f528bda36aa38a4d9ff5915" => :high_sierra
+  end
+
+  head do
```
Comment (DomT4) [inline Formula/go@1.9.rb]: `head` block should be pinned to the 1.9 branch or removed entirely.

Comment (fxcoudert) [reply Formula/go@1.9.rb]: Removed is good

File: Formula/go@1.9.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    url "https://go.googlesource.com/go.git"
+
+    resource "gotools" do
+      url "https://go.googlesource.com/tools.git"
+    end
+  end
+
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: Here: `keg_only :versioned_formula`

File: Formula/go@1.9.rb (hunk 1 above)
Comment (mkabischev) [reply Formula/go@1.9.rb]: removed

File: Formula/go@1.9.rb (hunk 2 above)
Comment (mkabischev) [reply Formula/go@1.9.rb]: removed

File: Formula/go@1.9.rb (hunk 3 above)
Comment (mkabischev) [reply Formula/go@1.9.rb]: added

File: Formula/go@1.9.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
+class GoAT19 < Formula
+  desc "Go programming environment (1.9)"
+  homepage "https://golang.org"
+
+  stable do
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: As there is no head block anymore, `stable do` should be removed and the content of the block left-shifted one level.

File: Formula/go@1.9.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    url "https://dl.google.com/go/go1.9.4.src.tar.gz"
+    mirror "https://fossies.org/linux/misc/go1.9.4.src.tar.gz"
+    sha256 "0573a8df33168977185aa44173305e5a0450f55213600e94541604b75d46dc06"
+
+    go_version = version.to_s.split(".")[0..1].join(".")
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: `go_version` will forever be `1.9`, so simply remove the variable and write `1.9` below

File: Formula/go@1.9.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    resource "gotools" do
+      url "https://go.googlesource.com/tools.git",
+          :branch => "release-branch.go#{go_version}"
+    end
+  end
+
+  keg_only :versioned_formula
+
+  option "without-cgo", "Build without cgo (also disables race detector)"
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: Analytics show that these two options are used in 0.01% of go installs, and are never ever used in versioned go formulas such as go 1.8. Let's drop the options.

File: Formula/go@1.9.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  def install
+    (buildpath/"gobootstrap").install resource("gobootstrap")
+    ENV["GOROOT_BOOTSTRAP"] = buildpath/"gobootstrap"
+
+    cd "src" do
+      ENV["GOROOT_FINAL"] = libexec
+      ENV["GOOS"]         = "darwin"
+      ENV["CGO_ENABLED"]  = "0" if build.without?("cgo")
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: Remove that line, as the option is gone

File: Formula/go@1.9.rb [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    end
+
+    (buildpath/"pkg/obj").rmtree
+    rm_rf "gobootstrap" # Bootstrap not required beyond compile.
+    libexec.install Dir["*"]
+    bin.install_symlink Dir[libexec/"bin/go*"]
+
+    # Race detector only supported on amd64 platforms.
+    # https://golang.org/doc/articles/race_detector.html
+    if build.with?("cgo") && build.with?("race") && MacOS.prefer_64_bit?
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: Supported platforms are all 64-bit, so remove the `if` and run the next line unconditionally.

File: Formula/go@1.9.rb [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    # Run go fmt check for no errors then run the program.
+    # This is a a bare minimum of go working as it uses fmt, build, and run.
+    system bin/"go", "fmt", "hello.go"
+    assert_equal "Hello World\n", shell_output("#{bin}/go run hello.go")
+
+    # godoc was installed
+    assert_predicate libexec/"bin/godoc", :exist?
+    assert_predicate libexec/"bin/godoc", :executable?
+
+    if build.with? "cgo"
```
Comment (fxcoudert) [inline Formula/go@1.9.rb]: Also, remove the `if`

**Review by fxcoudert — DISMISSED** [review]

**Comment by ilovezfs** [comment]
Thanks for your first contribution to Homebrew, @mkabischev! Without people like you submitting PRs we couldn't run this project. You rock!

When you have a chance, please make sure the email address you use for commits is actually associated with your GitHub account so that GitHub contribution statistics are correct.

**Pull request merged**